File: homework/hardware_software/homework5/51215902152-侯雪城-第五次作业/4.2题源码/排序源码/QuickSort/quick_sort.c
#include"quick_sort.h"

struct node
{
	uint8 low,high;
};

uint8 partition(uint8 *arr,uint8 low,uint8 high)
{
	uint8 base = arr[low];
	while (low<high)
	{
		while (low<high && arr[high]>=base)
		{
			high--;
		}
		arr[low] = arr[high];
		while (low<high && arr[low]<=base)
		{
			low++;
		}
		arr[high] = arr[low];
	}
	arr[low] = base;
	return low;
}

void quickSort(uint8 *arr,uint8 low,uint8 high)
{
	struct node st[500];
	st[0].low=low;
	st[0].high=high;
	int8 top = 0;
	while(true)
	{
		if (top<0) break;
		low = st[top].low;
		high = st[top].high;
		top = top - 1;
		if (low < high)
		{
			uint8 base = partition(arr,low,high);

			if (base > low)
			{
				top++;
				st[top].low = low;
				st[top].high = base - 1;
			}
			if (base < high)
			{
				top++;
				st[top].low = base + 1;
				st[top].high = high;

			}
		}
	}
}

File: homework/hardware_software/homework5/51215902152-侯雪城-第五次作业/4.1题源码/Matlab源码/排序源码/quick_sort.m
function s=quick_sort(s)
    if(length(s)>1)
        i=1;
        j=length(s);
        x=s(1);
        while(i<j)
            while(i<j&&s(j)>x)
                j=j-1;
            end
            if(i<j)
                s(i)=s(j);
                i=i+1;
            end
            while(i<j&&s(i)<x)
                i=i+1;
            end
            if(i<j)
                s(j)=s(i);
                j=j-1;
            end
        end
        s(i)=x;
        if (i>1)
            s(1:i-1)=quick_sort(s(1:i-1));
        end
        if (i<length(s))
            s(i+1:end)=quick_sort(s(i+1:end));
        end
    end
end

        
        